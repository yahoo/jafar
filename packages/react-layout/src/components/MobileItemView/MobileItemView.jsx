/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { useState } from 'react';
import PropTypes from 'prop-types';
import IconButton from '@material-ui/core/IconButton';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import Section from '../Section';
import Menu from '../Menu';
import { ITEM_HEIGHT } from '../Menu/Menu';
import Style from './Styled';

const menuRef = React.createRef();
const elementRef = React.createRef();

function MobileItemView(props) {
  const [isMenuOpen, setIsMenuOpen] = useState(false);

  const openMenu = () => {
    setIsMenuOpen(true);
  };

  const closeMenu = () => {
    setIsMenuOpen(false);
  };

  const Styled = Style(props.size);

  return (<Styled.Wrapper ref={elementRef}>
    {
      (props.title || props.mainActions || props.optionsActions) && <Styled.Header>
        {
          props.title
            && <Styled.Title>{props.title}</Styled.Title>
        }
        {
          props.mainActions && <span>{ props.mainActions
            .filter(action => action.icon)
            .map((action, index) => {
              const Icon = action.icon;
              const key = action.label || index;
              return (<IconButton key={key} aria-label={key}
                color="primary" disabled={action.disabled} onClick={action.onClick}><Icon /></IconButton>)
              ;
            }) }</span>
        }
        {
          props.optionsActions
            && <React.Fragment>
              <IconButton color="primary" onClick={openMenu} aria-label="Options"><MoreVertIcon /></IconButton>
              <Menu options={props.optionsActions} anchorEl={menuRef.current} open={isMenuOpen}
                onClose={closeMenu}
                getContentAnchorEl={null}
                anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
                transformOrigin={{ vertical: 'top', horizontal: 'center' }}
                style={{
                  height: `${(props.optionsActions.length * ITEM_HEIGHT) + 10}px`,
                  maxHeight: `${((elementRef.current || {}).clientWidth - 100)}px`,
                  width: `${((elementRef.current || {}).clientWidth - 40)}px`,
                  marginTop: '-20px',
                }} />
            </React.Fragment>
        }
      </Styled.Header>
    }
    <Styled.Sections aria-label="Sections">
      {
        props.sections.map((section, index) => (<Section key={section.id} size={props.size} {...section} />))
      }
    </Styled.Sections>
    <div id="menu-anchor" ref={menuRef} />
  </Styled.Wrapper>);
}

MobileItemView.propTypes = {
  title: PropTypes.string,
  sections: PropTypes.array.isRequired,
  mainActions: PropTypes.array,
  optionsActions: PropTypes.array,
  size: PropTypes.number,
};

MobileItemView.defaultProps = {
  size: 1,
};

export default MobileItemView;
