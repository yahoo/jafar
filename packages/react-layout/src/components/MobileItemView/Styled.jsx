/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import styled from 'styled-components';

const mobileItemViewSizeCache = {};

export default (size) => {
  const key = size;
  if (!mobileItemViewSizeCache[key]) {
    mobileItemViewSizeCache[key] = {
      Wrapper: styled.div`
        color: #3d3d3d;
        background-color: #fff;
        height:100%;
        display: flex;
        flex-direction: column;
      `,
      Header: styled.div`
        background-color: #f0f0f0;
        font-size: 16px;
        padding: 10px 10px 10px ${10 + (size * 10)}px;
        display: flex;
        flex-direction: row;
        flex: 0 0 auto;
      `,
      Title: styled.div`
        font-size: 18px;
        font-weight: bold;
        flex: 1;
        line-height: 45px;
      `,
      Sections: styled.div`
        flex: 1;
        overflow-y: auto;
      `,
    };
  }
  return mobileItemViewSizeCache[key];
};

