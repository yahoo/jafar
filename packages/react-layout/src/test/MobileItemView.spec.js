/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import MobileItemView from '../components/MobileItemView/index';

describe('MobileItemView', () => {
  const optionsSelector = 'aria-label="Options"';
  let props;

  beforeEach(() => {
    props = {
      title: 'Employee',
      sections: [{
        id: 'general',
        title: 'General',
      }],
      mainActions: [{
        label: 'Save',
        icon: () => <svg />,
        type: 'primary',
        onClick: jest.fn(),
      }],
      optionsActions: [{
        label: 'History',
        onClick: jest.fn(),
      }],
    };
  });

  it('Should render - ok', () => {
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });

  it('Should render without title - ok', () => {
    props.title = undefined;
    const component = shallow(getComponent(props));
    const html = component.html();
    expect(html).toContain(optionsSelector);
    expect(html).not.toContain(props.title);
    expect(html).toContain(props.mainActions[0].label);
  });

  it('Should render without mainActions - ok', () => {
    const mainActionIcon = props.mainActions[0].label;
    props.mainActions = undefined;
    const component = shallow(getComponent(props));
    const html = component.html();
    expect(html).toContain(optionsSelector);
    expect(html).toContain(props.title);
    expect(html).not.toContain(mainActionIcon);
  });

  it('Should render without optionsActions - ok', () => {
    props.optionsActions = undefined;
    const component = shallow(getComponent(props));
    const html = component.html();
    expect(html).not.toContain(optionsSelector);
    expect(html).toContain(props.title);
    expect(html).toContain(props.mainActions[0].label);
  });

  it('Should render with only optionsActions - ok', () => {
    const mainActionIcon = props.mainActions[0].label;
    props.title = undefined;
    props.mainActions = undefined;
    const component = shallow(getComponent(props));
    const html = component.html();
    expect(html).toContain(optionsSelector);
    expect(html).not.toContain(props.title);
    expect(html).not.toContain(mainActionIcon);
  });

  function getComponent(props) {
    return (<MobileItemView {...props} />);
  }
});
