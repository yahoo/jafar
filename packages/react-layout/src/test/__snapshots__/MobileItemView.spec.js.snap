// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileItemView Should render - ok 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      Employee
    </styled.div>
    <span>
      <WithStyles(ForwardRef(IconButton))
        aria-label="Save"
        color="primary"
        key="Save"
        onClick={[MockFunction]}
      >
        <icon />
      </WithStyles(ForwardRef(IconButton))>
    </span>
    <WithStyles(ForwardRef(IconButton))
      aria-label="Options"
      color="primary"
      onClick={[Function]}
    >
      <MoreVertIcon />
    </WithStyles(ForwardRef(IconButton))>
    <Menu
      anchorEl={null}
      anchorOrigin={
        Object {
          "horizontal": "center",
          "vertical": "bottom",
        }
      }
      getContentAnchorEl={null}
      onClose={[Function]}
      open={false}
      options={
        Array [
          Object {
            "label": "History",
            "onClick": [MockFunction],
          },
        ]
      }
      style={
        Object {
          "height": "58px",
          "marginTop": "-20px",
          "maxHeight": "NaNpx",
          "width": "NaNpx",
        }
      }
      transformOrigin={
        Object {
          "horizontal": "center",
          "vertical": "top",
        }
      }
    />
  </styled.div>
  <styled.div
    aria-label="Sections"
  >
    <ForwardRef
      id="general"
      key="general"
      size={1}
      title="General"
    />
  </styled.div>
  <div
    id="menu-anchor"
  />
</styled.div>
`;
